File: app/src/main/java/com/lahtinen/apps/flextime/Calculator.java
package com.lahtinen.apps.flextime;

import java.util.ArrayList;
import java.util.Collections;
import java.util.concurrent.TimeUnit;

import static java.lang.Integer.parseInt;

public class Calculator {

    public static String intMinutesToString(int minutes) {
        if (minutes == 0) {
            return "0h";
        }
        final long hourPart = TimeUnit.MINUTES.toHours(minutes);
        final long minutePart = minutes - TimeUnit.HOURS.toMinutes(TimeUnit.MINUTES.toHours(minutes));
        return (hourPart != 0 ? hourPart + "h" : "") + (minutePart != 0 ? minutePart + "m" : "");
    }

    public static int stringToInt(String timeString) {
        final String[] times = timeString.split(":");
        return parseInt(times[0]) * 60 + parseInt(times[1]);
    }

    public static String[] buildPickerNumbers() {
        final String[] hourValues = new String[]{"00", "01", "02", "03", "04"};
        final String[] minuteValues = new String[]{"00", "15", "30", "45"};

        final ArrayList<String> times = buildTime(hourValues, minuteValues);
        Collections.reverse(times);

        final String[] timesArray = new String[times.size()];
        return times.toArray(timesArray);
    }

    private static ArrayList<String> buildTime(String[] hourValues, String[] minuteValues) {
        final ArrayList<String> times = new ArrayList<>();
        for (String hour : hourValues) {
            for (String minute : minuteValues) {
                times.add(hour + ":" + minute);
            }
        }
        return times;
    }
}

File: app/src/main/java/com/lahtinen/apps/flextime/Calculator.kt
package com.lahtinen.apps.flextime

import java.lang.Integer.parseInt

class Calculator {

    companion object {
        fun timeStringToMinutesInt(timeString: String): Int {
            val times = timeString.split(":")
            val hours = parseInt(times[0])
            val minutes = parseInt(times[1])
            return hours * 60 + if (timeString[0] == '-') -minutes else minutes
        }

        fun buildPickerTimes(): Array<String> {
            val hours = listOf("00", "01", "02", "03", "04")
            val minutes = listOf("00", "15", "30", "45")

            val times = buildTime(hours, minutes)
            return times.reversed().toTypedArray()
        }

        private fun buildTime(hourValues: List<String>, minuteValues: List<String>): List<String> {
            return hourValues.flatMap { hour -> minuteValues.map { minute -> "$hour:$minute" } }
        }
    }
}
